dynamiqs/integrators/core/rouchon_integrator.py

```diff
@@ -208,3 +214,75 @@ def _kraus_ops(self, t: float, dt: float) -> list[QArray]:
 mesolve_rouchon3_integrator_constructor = lambda **kwargs: MESolveRouchon3Integrator(
     **kwargs, diffrax_solver=RouchonDXSolver(3), fixed_step=True
 )
+
+
+class MESolveAdaptiveRouchon12Integrator(MESolveDiffraxIntegrator):
+    """Integrator computing the time evolution of the Lindblad master equation using the
+    adaptive Rouchon 1-2 method.
+    """
+
+    @property
+    def terms(self) -> dx.AbstractTerm:
+        def kraus_map(t0, t1, y0):  # noqa: ANN202
+            rho = y0
+            t = (t0 + t1) / 2
+            dt = t1 - t0
+
+            L, H = self.L(t), self.H(t)
+
+            # === first order
+            Ms_1 = MESolveRouchon1Integrator.Ms(H, L, dt, self.method.exact_expm)
+            rho_1 = cholesky_normalize(Ms_1, rho) if self.method.normalize else rho
+            rho_1 = sum([M @ rho_1 @ M.dag() for M in Ms_1])
+
+            # === second order
+            Ms_2 = MESolveRouchon2Integrator.Ms(H, L, dt, self.method.exact_expm)
+            rho_2 = cholesky_normalize(Ms_2, rho) if self.method.normalize else rho
+            rho_2 = sum([M @ rho_2 @ M.dag() for M in Ms_2])
+
+            return rho_2, rho_2 - rho_1
```

Comment:
Note that Patrick uses `0.5 * (rho_2 - rho_1)` for the error estimate in diffrax. See e.g. https://github.com/patrick-kidger/diffrax/blob/d3c1430b76e7d0158536823724371d7b26f5db8f/diffrax/_solver/implicit_euler.py#L96. Of course, it will only change how restrictive or not you are with `dt`, so I don't think it really matters.